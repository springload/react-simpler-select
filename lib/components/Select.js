import React from 'react';
import PropTypes from 'prop-types';

// Converts a DOM NodeList or other iterable to a JS array.
const toArray = iterable => Array.prototype.slice.call(iterable);

// Filters props to remove before transmission to the select element.
const filterProps = (props) => {
  return Object.keys(props)
      .filter(k => ['onChange', 'options', 'placeholder'].indexOf(k) === -1)
      .reduce((filtered, k) => {
        // eslint-disable-next-line no-param-reassign
        filtered[k] = props[k];
        return filtered;
      }, {});
};

/**
 * The Select component. Renders to a plain HTML <select>.
 */

const propTypes = {
  onChange: PropTypes.func.isRequired,
  options: PropTypes.array.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.any,
};

class Select extends React.Component {
  handleChange(e) {
    let value;

    if (this.isMultiSelect()) {
      value = toArray(e.target.options)
          .filter(option => option.selected)
          .map(option => option.value);
    } else {
      value = e.target.value;
    }

    this.props.onChange(value);
  }

  isMultiSelect() {
    const { value } = this.props;

    return Array.isArray(value);
  }

  renderOption(option) {
    return (
      <option
        key={option.value}
        value={option.value}
        disabled={option.disabled}
      >
        {option.label}
      </option>
    );
  }

  renderChildren() {
    const { options, placeholder } = this.props;

    const children = options.map((item) => {
      let child;

      if (item.options) {
        child = (
          <optgroup label={item.title} key={item.id}>
            {item.options.map(this.renderOption)}
          </optgroup>
        );
      } else {
        child = this.renderOption(item);
      }

      return child;
    });

    if (placeholder) {
      children.unshift(this.renderOption({ label: placeholder, value: '', disabled: true }));
    }

    return children;
  }

  render() {
    return (
      <select
        {...filterProps(this.props)}
        onChange={(e) => this.handleChange(e)}
        multiple={this.isMultiSelect()}
      >
        {this.renderChildren()}
      </select>
    );
  }
}

Select.propTypes = propTypes;

export default Select;
