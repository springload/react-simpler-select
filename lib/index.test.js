import React from 'react';
import { shallow } from 'enzyme';
import Select from './index';

// Converts a DOM NodeList or other iterable to a JS array.
const toArray = iterable => Array.prototype.slice.call(iterable);

const options = [
  { value: 'es', label: 'Spanish' },
  { value: 'cs', label: 'Czech' },
  { value: 'en', label: 'English' },
  { value: 'fi', label: 'Finnish' },
  { value: 'fr', label: 'French' },
];

const optgroups = [
  {
    title: 'North Island',
    id: 'ni',
    options: [
      { value: 'wgtn', label: 'Wellington' },
      { value: 'gsb', label: 'Gisbourne' },
      { value: 'oh', label: 'Ohakune' },
    ],
  },
  {
    title: 'South Island',
    id: 'si',
    options: [
      { value: 'ch', label: 'Christchurch' },
      { value: 'qt', label: 'Queenstown' },
      { value: 'te', label: 'lake Tekapo' },
    ],
  },
];

describe('Select', () => {
  it('renders', () => {
    expect(shallow((
      <Select
        name="decision"
        id="decision-selector"
        className="field-select"
        placeholder="Your decision"
        options={[]}
        required
      />
    ))).toMatchSnapshot();
  });

  it('#options', () => {
    expect(shallow((
      <Select
        options={options}
        value="en"
      />
    ))).toMatchSnapshot();
  });

  it('#onChange', () => {
    const handleChange = jest.fn();
    const wrapper = shallow((
      <Select
        options={options}
        onChange={handleChange}
        value="en"
      />
    ));

    wrapper.find('select').simulate('change', {
      target: { value: 'fi' }
    });

    expect(handleChange).toHaveBeenCalledWith('fi');
  });
});
