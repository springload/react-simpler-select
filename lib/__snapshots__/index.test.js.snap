// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select #options 1`] = `
<select
  multiple={false}
  onChange={[Function]}
  value="en"
>
  <option
    value="es"
  >
    Spanish
  </option>
  <option
    value="cs"
  >
    Czech
  </option>
  <option
    value="en"
  >
    English
  </option>
  <option
    value="fi"
  >
    Finnish
  </option>
  <option
    value="fr"
  >
    French
  </option>
</select>
`;

exports[`Select renders 1`] = `
<select
  className="field-select"
  id="decision-selector"
  multiple={false}
  name="decision"
  onChange={[Function]}
  required={true}
>
  <option
    disabled={true}
    value=""
  >
    Your decision
  </option>
</select>
`;
